lib/membrane/rtp/vad_utils/is_speaking_estimator.ex

```diff
@@ -0,0 +1,118 @@
+
+  defp binomial_coefficient(n, k) do
+    k = min(k, n - k)
+
+    Enum.to_list(1..k)
+    |> Enum.reduce(1, fn i, acc -> div(acc * (n - i + 1), i) end)
+  end
+
+  defp compute_activity_score(vl, n_r, lambda) do
+    p = 0.5
+
+    score =
+      :math.log(binomial_coefficient(n_r, vl)) +
+        vl * :math.log(p) +
+        (n_r - vl) * :math.log(1 - p) -
+        :math.log(lambda) + lambda * vl
+
+    if score > @min_activity_score, do: score, else: @min_activity_score
+  end
+
+  defp indicator_func(x, threshold) do
+    if x >= threshold, do: 1, else: 0
+  end
+
+  defp count_active(data, threshold) do
+    Enum.reduce(data, 0, fn x, acc -> acc + indicator_func(x, threshold) end)
+  end
+
+  defp compute_interval(littles, sub_list_length, threshold) do
+    littles
+    |> Enum.chunk_every(sub_list_length)
+    |> Enum.map(&count_active(&1, threshold))
+  end
+
+  defp compute_immediates(levels, level_threshold),
+    do: compute_interval(levels, @n1, level_threshold)
+
+  defp compute_mediums(immediates), do: compute_interval(immediates, @n2, @medium_threshold)
+  defp compute_longs(mediums), do: compute_interval(mediums, @n3, @long_threshold)
+
+  defp scores_above_threshold?(immediate_score, medium_score, long_score),
+    do:
+      immediate_score > @immediate_score_threshold and
+        medium_score > @medium_score_threshold and
+        long_score > @long_score_threshold
+
+  @spec estimate_is_speaking(list(integer), integer) :: :speech | :silence
+  def estimate_is_speaking(levels, _level_threshold) when length(levels) < @min_levels_length,
+    do: :silence
+
+  def estimate_is_speaking(levels, level_threshold) do
+    immediates = compute_immediates(levels, level_threshold)
+    mediums = compute_mediums(immediates)
+    longs = compute_longs(mediums)
+
+    immediate_score = immediates |> hd() |> compute_activity_score(@n1, 1)
+    medium_score = mediums |> hd() |> compute_activity_score(@n2, 24)
+    long_score = longs |> hd() |> compute_activity_score(@n3, 47)
+
+    above_threshold? = scores_above_threshold?(immediate_score, medium_score, long_score)
+
+    if above_threshold?,
+      do: :speech,
+      else: :silence
+  end
+
+  @spec safe_trim_queue(Qex.t(), integer) :: Qex.t()
+  defp safe_trim_queue(queue, n) do
+    if Enum.count(queue) > @min_levels_length do
+      {trimmed_queue, _rest} = Qex.split(queue, n)
+      trimmed_queue
+    else
+      queue
+    end
+  end
+
+  # Takes the queue from RTP VAD module and returns the queue of `@min_levels_length` length and the estimation
+  @spec trim_queue_and_estimate_vad(Qex.t(), integer) :: {Qex.t(), :speech | :silence}
+  def trim_queue_and_estimate_vad(queue, threshold) do
+    trimmed_queue = safe_trim_queue(queue, @min_levels_length)
+
+    estimation =
+      trimmed_queue
+      |> Enum.map(fn {level, _timestamp} -> 127 + level end)
+      |> estimate_is_speaking(threshold)
+
+    {trimmed_queue, estimation}
+  end
```

Comment:
This is a bit confusing when it comes to separating the responsibilities. Is the queue internal to the `IsSpeakingEstimator` module or not? If yes, then it should be responsible also for putting the audio levels inside (BTW it could then apply the offset as now you're mapping the entries on each call). Otherwise, it should not be responsible for trimming the queue.

Reply:
I'll make `vad.ex` responsible for the queue and change the range of values inside of it from [-127, 0] to [0, 127]. Then I'll change the module of the given function.